src/sequential/sequential.cpp

```diff
@@ -0,0 +1,275 @@
+#include "sequential.hpp"
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// HELPER CLASSES
+////////////////////////////////////////////////////////////////////////////////
+
+bool
+SequentialBucket::is_empty() {
+  LOG_TRACE("Enter");
+  return this->offset == SIZE_MAX;
+}
+
+void
+SequentialBucket::set(const KeyType key,
+                      const ValueType value,
+                      const HashCodeType hashcode,
+                      const size_t offset) {
+  LOG_TRACE("Enter");
+  assert(this->is_empty() && "overwriting non-empty SequentialBucket");
+  this->key = key;
+  this->value = value;
+  this->hashcode = hashcode;
+  this->offset = offset;
+}
+
+void
+SequentialBucket::swap(SequentialBucket &other) {
+  LOG_TRACE("Enter");
+  KeyType tmp_key = this->key;
+  this->key = other.key;
+  other.key = tmp_key;
+
+  ValueType tmp_value = this->value;
+  this->value = other.value;
+  other.value = tmp_value;
+
+  HashCodeType tmp_hc = this->hashcode;
+  this->hashcode = other.hashcode;
+  other.hashcode = tmp_hc;
+
+  size_t tmp_offset = this->offset;
+  this->offset = other.offset;
+  other.offset = tmp_offset;
+}
+
+bool
+SequentialBucket::equal_key(const KeyType key, const HashCodeType hashcode) {
+  LOG_TRACE("Enter");
+  assert(!this->is_empty() && "comparing to empty bucket!");
+  // Assume equality between hashcodes is simpler (because it is fixed for any
+  // type of key)
+  return this->hashcode == hashcode && this->key == key;
+}
+
+void
+SequentialBucket::print(const size_t size) {
+  if (this->is_empty()) {
+    std::cout << "(empty)";
+  } else {
+    std::cout << "(" << this->hashcode << "=>" << this->hashcode % size <<
+        "+" << this->offset << ") " << this->key << ": " << this->value;
+  }
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// STATIC HELPER FUNCTIONS
+////////////////////////////////////////////////////////////////////////////////
+
+HashCodeType
+hash(const KeyType key) {
+  LOG_TRACE("Enter");
+  size_t k = static_cast<size_t>(key);
+  // I use the suffix '*ULL' to denote that the literal is an int64. I've had
+  // weird bugs in the past to do with literal conversion. I'm not sure the
+  // details. I only remember it was a right pain.
+  k = ((k >> 30) ^ k) * 0xbf58476d1ce4e5b9ULL;
+  k = ((k >> 27) ^ k) * 0x94d049bb133111ebULL;
+  k =  (k >> 31) ^ k;
+  return reinterpret_cast<HashCodeType>(k);
+}
+
+size_t
+get_home(const HashCodeType hashcode, const size_t size) {
+  LOG_TRACE("Enter");
+  size_t h = reinterpret_cast<HashCodeType>(hashcode);
+  return h % size;
+}
+
+/// @brief  Get real bucket index.
+size_t
+get_probe(const size_t home, const size_t offset, const size_t size) {
+  LOG_TRACE("Enter");
+  return (home + offset) % size;
+}
+
+std::pair<SearchStatus, size_t>
+get_wouldbe_offset(      std::vector<SequentialBucket> &tmp_buckets,
```

Comment:
Follow up question. What does a `const` reference mean? Does it mean the pointer won't change or that the data pointed to won't change? Or both?

It's just so clear with C pointers:

* `const int *x` is a variable pointer to an array of constant integers
* `int *const x` is a constant pointer to an array of variable integers

Reply:
References in C++ have the same properties as pointers. The only difference is that they are bound once, and can never be null. This means if you see `const` with a reference, it always means the pointee cannot be modified by the reference holder. I.e., `const T &x` means reference to constant T.

Some more background - const in C/C++ is confusing in general because it applies to the thing to its left, UNLESS there is nothing to its left. So it would actually be way clearer if everybody just stuck to that convention:
```c++
int const *x; // Pointer to constant integer.
int *const x; // Constant pointer to integer.
int const *const x; // Constant pointer to constant integer.
```
I believe this is called "east-const". I've considered adopting it for clarity, but realised a lot of people still find it unexpected.

Reply:
I agree that your convention is clearer. I learned to read types from "right-to-left". I think it's still less common to do `int const x;` than `const in x;`, which is why I do it the way I do. But for this project, I can switch. And in the future, I may switch anyhow too.

Reply:
I'll switch if you switch.

Reply:
Lol what are _you_ switching to? 😆

Reply:
I'm still doing west const because that seems to be status quo, but let's switch to east const since it's unambiguous.